routing/DELEGATED_CONTENT_ROUTING_HTTP.md

```diff
@@ -0,0 +1,99 @@
+
+**Author(s)**:
+- Gus Eggert
+
+**Maintainer(s)**:
+
+* * *
+
+**Abstract**
+
+"Delegated content routing" is a mechanism for IPFS implementations to use for offloading content routing to another process/server. This spec describes an HTTP API for delegated routing.
+
+# API Specification
+The Delegated Content Routing Routing HTTP API uses the `application/json` content type by default. Clients and servers *should* support `application/cbor`, which can be negotiated using the standard `Accept` and `Content-Type` headers.
+
+## Common Data Types:
+
+- CIDs are always encoded using a [multibase](https://github.com/multiformats/multibase)-encoded [CIDv1](https://github.com/multiformats/cid#cidv1).
+- Multiaddrs are encoded according to the [human-readable multiaddr specification](https://github.com/multiformats/multiaddr#specification)
+- Peer IDs are encoded according [PeerID string representation specification](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#string-representation)
+- Multibase bytes are encoded according to [the Multibase spec](https://github.com/multiformats/multibase), and *should* use Base64.
+- Timestamps are Unix millisecond epoch timestamps
+
+## API
+- `GET /v1/providers/{CID}`
+    - Response
+        
+        ```json
+        {
+            "Providers": [
+                {
+					"Peer": {
+						"ID": "12D3K...",
+						"Multiaddrs": ["/ip4/.../tcp/.../p2p/...", "/ip4/..."]
+					}
+                    "Protocols": [
+                        {
+                            "Codec": 2320,
+                            "Payload": { ... }
+                        }
+                    ]
+                }
+            ]
+        }
+        ```
+        
+    - Default limit: 100 providers
+    - Optional query parameters
+        - `transfer` only return providers who support the passed transfer protocols, expressed as a comma-separated list of [multicodec codes](https://github.com/multiformats/multicodec/blob/master/table.csv) in decimal form such as `2304,2320`
+        - `transport` only return providers whose published multiaddrs explicitly support the passed transport protocols, such as `460,478` (`/quic` and `/tls/ws`)
+        - Servers should treat the multicodec codes used in the `transfer` and `transport` parameters as opaque, and not validate them, for forwards compatibility
+- `GET /v1/providers/hashed/{multihash}`
+    - This is the same as `GET /v1/providers/{CID}`, but takes a hashed CID encoded as a [multihash](https://github.com/multiformats/multihash/)
+- `PUT /v1/providers`
+    - Reframe equivalent: Provide
+    - Body
+        ```json
+        {
+            "Signature": "multibase bytes",
+            "Payload": "multibase bytes"
+        }
+        ```
+	    - `Payload` is a multibase-encoded string containing a JSON object with the following schema:
+		```json
+		{
+                "Keys": ["cid1", "cid2"],
```

Comment:
What would be the strategy for supporting large nodes with >1B cids? Pagination could be a useful feature. Maybe we can have something like that in the protocol spec:

```
"Page": 2.
"PagesTotal": 10
```

Knowing the current and total page numbers would help to understand where the snapshot ends and would help to determine what CIDs have expired.

Reply:
You can always provide 1B CIDs doing several calls to the server, right?

Reply:
If you want to avoid early expiration from a CID, you can always send the DAG from bottom to top, so the indexer will have an earlier TTL on the root CID, avoiding the expiration of the lowest part of the tree, right? 

In any case, how can `Pages` and `PagesTotal` give you any context about the snapshot you want to provide? Maybe we could add the `RootCID` to the query if you need some context about the `PUT provide` request.

Reply:
From the indexer's sidecar POV we'd like to consume the snapshot in chunks and not as a single blob, due to RAM implications, hence the paging suggestion above.

If a snapshot would come in chunks then the indexer's sidecar would also need to know where the snapshot ends so that it can determine what CIDs have expired since the last snapshot.

Reply:
I'm a bit confused here, it doesn't make much sense to paginate a PUT request. Are you referring to the GET request?

If so, I would like to avoid PagesTotal as that can be expensive to determine depending on the database implementation and size.

If your goal is to determine the CIDs that have expired since some time, would a filter on TTL work?

Reply:
This thread can be resolved now:
1. [PR](https://github.com/ipfs/go-delegated-routing/pull/64) for enabling calling `Provide` endpoint in batches
2. [Issue](https://github.com/ipfs/specs/issues/346) for implementing `Delete` endpoint for explicitly flagging what has been deleted, as well as smarter tracking of CID deltas.